### PR TITLE
Introspection Test Api

### PR DESCRIPTION
**Pull request by do4gr**

### CODE REVIEW

File: introspection-engine/introspection-engine-tests/tests/add_prisma1_defaults/mod.rs [hunk 1]
```diff
@@ -22,13 +22,19 @@ async fn add_cuid_default(api: &TestApi) -> crate::TestResult {
...
-        }
-    "#};
+    let native_type = if sql_family.is_postgres() {
+        "@db.VarChar(25)"
+    } else {
+        "@db.Char(25)"
+    };
 
-    assert_eq_datamodels!(dm, &api.introspect().await?);
+    let dm = formatdoc! {r#"
```
Comment (do4gr) [inline introspection-engine/introspection-engine-tests/tests/add_prisma1_defaults/mod.rs:31]: Thought from discussion with Tom: having a barrel extension where type::p_string has our default mappings for native types for the Prisma type string. This way we avoid having to sprinkle native types everywhere when that is not the test focus.

File: introspection-engine/introspection-engine-tests/tests/tables/mod.rs [hunk 2]
```diff
@@ -640,19 +671,60 @@ async fn negative_default_values_should_work(api: &TestApi) -> crate::TestResult
...
-    "##};
+        }}
+    "##, float_native, float_native};
 
-    assert_eq_datamodels!(dm, &api.introspect().await?);
+    api.assert_eq_datamodels(&dm, &api.introspect().await?);
 
     Ok(())
 }
+
```
Comment (do4gr) [inline introspection-engine/introspection-engine-tests/tests/tables/mod.rs:698]: This was for the bug I initially intended to fix. Will pick this up after merging.

File: introspection-engine/introspection-engine-tests/src/test_api.rs [hunk 3]
```diff
@@ -183,6 +192,42 @@ impl TestApi {
...
+    }
+
+    pub fn dm_with_sources(&self, schema: &str) -> String {
+        let mut out = String::with_capacity(320 + schema.len());
+
+        out.push_str(&self.datasource_block());
+        out.push_str(schema);
+
+        out
+    }
```
Comment (tomhoule) [inline introspection-engine/introspection-engine-tests/src/test_api.rs:230]: Much better do include the datasource in the assertions like this :+1:

**Review by tomhoule — APPROVED** [review]
Very neat!

**Pull request merged**